### PR TITLE
EnableScoring::Disabled - optional Searcher

### PR DESCRIPTION
**Pull request by shikhar**
We have a custom `Query` that requires the `Searcher` in `weight(..)` (for looking up cached data used in matching that is kept warmed and keyed on searcher generation), even when scoring is disabled (e.g. using `DocSetCollector`)

So I am proposing to make a further API change on top of https://github.com/quickwit-oss/tantivy/pull/1667 such that `EnableScoring::Disabled` can contain an optional `Searcher`, and nudging towards providing it when available. 

`Searcher::search(..)` has been updated accordingly.

### CODE REVIEW

File: src/query/query.rs
Comment (fulmicoton) [inline src/query/query.rs]: nitpick

```suggestion
    Disabled {
            schema: &'a Schema, 
            searcher_opt: Option<&'a Searcher>),
      }
```

**Review by fulmicoton — APPROVED** [review]
approved but please consider addressing nitpick comment

**Comment by fulmicoton** [comment]
@shikhar you can hit the squash and merge button I think you have the permissions!

**Comment by shikhar** [comment]
yup was just waiting on CI :)

**Pull request merged**
